modules/login/login_module.e

```diff
@@ -0,0 +1,1037 @@
+			r.set_value ("Login", "optional_content_type")
+			r.execute
+		end
+
+	handle_workaround_filter (api: CMS_API; req: WSF_REQUEST; res: WSF_RESPONSE)
+		local
+			r: CMS_RESPONSE
+			br: BAD_REQUEST_ERROR_CMS_RESPONSE
+		do
+			create {GENERIC_VIEW_CMS_RESPONSE} r.make (req, res, api)
+			r.execute
+		end
+
+
+	handle_logout (api: CMS_API; req: WSF_REQUEST; res: WSF_RESPONSE)
+		local
+			r: CMS_RESPONSE
+			l_url: STRING
+			l_oauth_gmail: OAUTH_LOGIN
+			l_cookie: WSF_COOKIE
+		do
+			if
+				attached {WSF_STRING} req.cookie ({LOGIN_CONSTANTS}.oauth_session) as l_cookie_token and then
+				attached {CMS_USER} current_user (req) as l_user
+			then
+					-- Logout gmail
+				create l_cookie.make ({LOGIN_CONSTANTS}.oauth_session, l_cookie_token.value)
+				l_cookie.set_path ("/")
+				l_cookie.set_max_age (-1)
+				res.add_cookie (l_cookie)
+				unset_current_user (req)
+				create {GENERIC_VIEW_CMS_RESPONSE} r.make (req, res, api)
+				r.set_status_code ({HTTP_CONSTANTS}.found)
+				r.set_redirection (req.absolute_script_url (""))
+				r.execute
+			else
+				create {GENERIC_VIEW_CMS_RESPONSE} r.make (req, res, api)
+				r.set_status_code ({HTTP_CONSTANTS}.found)
+				l_url := req.absolute_script_url ("")
+				l_url.append ("/basic_auth_logoff")
+				r.set_redirection (l_url)
+				r.execute
+			end
+		end
+
+	handle_register (api: CMS_API; req: WSF_REQUEST; res: WSF_RESPONSE)
+		local
+			r: CMS_RESPONSE
+			l_user_api: CMS_USER_API
+			u: CMS_USER
+			l_roles: LIST [CMS_USER_ROLE]
+			l_exist: BOOLEAN
+			es: LOGIN_EMAIL_SERVICE
+			l_link: STRING
+			l_token: STRING
+			l_message: STRING
+		do
+			create {GENERIC_VIEW_CMS_RESPONSE} r.make (req, res, api)
+			r.set_value ("Register", "optional_content_type")
+			if req.is_post_request_method then
+				if
+					attached {WSF_STRING} req.form_parameter ("name") as l_name and then
+					attached {WSF_STRING} req.form_parameter ("password") as l_password and then
+					attached {WSF_STRING} req.form_parameter ("email") as l_email
+				then
+					l_user_api := api.user_api
+
+					if attached l_user_api.user_by_name (l_name.value) then
+							-- Username already exist.
+						r.values.force ("The user name exist!", "error_name")
+						l_exist := True
+					end
+					if attached l_user_api.user_by_email (l_email.value) then
+							-- Emails already exist.
+						r.values.force ("The email exist!", "error_email")
+						l_exist := True
+					end
+
+					if not l_exist then
+							-- New user
+						create {ARRAYED_LIST [CMS_USER_ROLE]}l_roles.make (1)
+						l_roles.force (l_user_api.authenticated_user_role)
+
+						create u.make (l_name.value)
+						u.set_email (l_email.value)
+						u.set_password (l_password.value)
+						u.set_roles (l_roles)
+						l_user_api.new_user (u)
+
+							-- Create activation token
+						l_token := new_token
+						l_user_api.new_activation (l_token, u.id)
+						create l_link.make_from_string (req.server_url)
```

Comment:
use l_url  instead of l_link
for me, "link" is html link ..
url is the location of html link.